### PR TITLE
En/review eg/augmented lagrangian

### PR DESCRIPTION
**Pull request by eduardolneto**
Mofiying augmented_lagrangian.py so that constraints comply with custom_losses structure.

### CODE REVIEW

File: essos/augmented_lagrangian.py [hunk 1]
```diff
@@ -15,150 +15,493 @@ class LagrangeMultiplier(NamedTuple):
...
+#     #     if model_lagrangian=='Standard' :
+#     #         return update_method(lagrange_params,updates,eta,omega,model_mu=model_mu,beta=beta,mu_max=mu_max,alpha=alpha,gamma=gamma,epsilon=epsilon,eta_tol=eta_tol,omega_tol=omega_tol), state
+#     #     elif model_lagrangian=='Squared' :
+#     #         return update_method_squared(lagrange_params,updates,eta,omega,model_mu=model_mu,beta=beta,mu_max=mu_max,alpha=alpha,gamma=gamma,epsilon=epsilon,eta_tol=eta_tol,omega_tol=omega_tol), state
+#     #     else:
+#     #         print('Lagrangian model not available please select Standard or Squared ')
+#     #         os._exit(0)              
+
+#     return optax.GradientTransformation(init_fn, update_fn)
+
```
Comment (EstevaoMGomes) [inline essos/augmented_lagrangian.py]: Please remove this commented section

**Review by EstevaoMGomes — COMMENTED** [review]
Overall looks good

File: essos/augmented_lagrangian.py (hunk 1 above)
Comment (eduardolneto) [reply essos/augmented_lagrangian.py]: @EstevaoMGomes  Removed this section. Also added the custom_losses compatible augmented_lagrangian coils vmec surface example. However, that example will need the additions to losses. py in pull request #39  to work properly.

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 2]
```diff
@@ -0,0 +1,249 @@
+import os
+number_of_processors_to_use = 1 # Parallelization, this should divide ntheta*nphi
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Perhaps parallelization is not needed here?

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+os.environ["XLA_FLAGS"] = f'--xla_force_host_platform_device_count={number_of_processors_to_use}'
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Same

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+
+init_curves = CreateEquallySpacedCurves(N_COILS, FOURIER_ORDER, LARGE_R, SMALL_R, n_segments=N_SEGMENTS, nfp=NFP, stellsym=STELLSYM)
+init_coils = Coils(curves=init_curves, currents=[COIL_CURRENT]*N_COILS)
+init_field = BiotSavart(init_coils)
+init_surface=surface
+
+""" Setting the losses weights and targets """
+LENGTH_WEIGHT = 1.; LENGTH_TARGET = 40.
+CURVATURE_WEIGHT = 1.; CURVATURE_TARGET = 0.5
+NORMAL_FIELD_WEIGHT = 1.
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: As in the other examples, the main parameters should have one line each, and be at the top before initializing other variables/classes

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+BdotN_Target_tol=1.e-6
+
+""" Creating the loss functions """
+def loss(field, surface):
+    return jnp.sum(jnp.abs(BdotN_over_B(surface, field)))
+
+def BdotN_constraint(field,surface,target_tol=1.e-6):
+    bdotn_over_b = BdotN_over_B(surface, field)
+    bdotn_over_b_loss = jnp.sqrt(jnp.sum(jnp.maximum(jnp.square(bdotn_over_b)-target_tol,0.0)))
+    #bdotn_over_b_loss = jnp.sqrt(jnp.sum(jnp.maximum((jnp.square(bdotn_over_b)-target_tol)/target_tol,0.0)))
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: This comment, and the ones below it, perhaps could be deleted to make the file simpler?

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    return bdotn_over_b_loss
+
+def loss_length_constraint(field):
+    return jnp.maximum(0, field.coils.length - LENGTH_TARGET)
+    #return jnp.maximum(0, (field.coils.length - LENGTH_TARGET)/LENGTH_TARGET)
+
+def loss_curvature_contraint(field):
+    return jnp.maximum(0, field.coils.curvature - CURVATURE_TARGET)
+    #return jnp.mean(jnp.maximum(0, (field.coils.curvature - CURVATURE_TARGET)/CURVATURE_TARGET))
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py:70]: same?

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+L_curvature_constraint.dependencies = {"field": init_field}
+L_normal_field_constraint.dependencies = {"field": init_field}
+L_length.dependencies = {"field": init_field}
+L_curvature.dependencies = {"field": init_field}
+
+
+
+
+
+
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Too many line breaks, can be deleted

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+mu_max=1.e4                                #Maximum penalty parameter allowed
+alpha=0.99                                  #These are parameters only used if gradient descent and adaaptative mu
+gamma=1.e-2
+epsilon=1.e-8
+omega_tol=1.e-7    #desired grad_tolerance, associated with grad of lagrangian to main parameters
+eta_tol=1.e-7    #desired contraint tolerance, associated with variation of contraints
+
+#curvature_constraint=alm.ScaledConstraint(alm.eq(loss_curvature_contraint,model_lagrangian=model_lagrangian, multiplier=multiplier,penalty=penalty,omega=omega,sq_grad=sq_grad))
+#length_constraint=alm.ScaledConstraint(alm.eq(loss_length_constraint,model_lagrangian=model_lagrangian, multiplier=multiplier,penalty=penalty,omega=omega,sq_grad=sq_grad))
+#field_constraint=alm.ScaledConstraint(alm.eq(BdotN_constraint,model_lagrangian=model_lagrangian, multiplier=multiplier,penalty=penalty,omega=omega,sq_grad=sq_grad))
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py:118]: If these comments are meant to show the user how to use different constraints/objectives, can you add a comment here saying that?

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+
+
+
+C_Total_constraint = alm.combine(C_normal_field_constraint, C_length_constraint, C_curvature_constraint)
+
+
+
+
+
+
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Again, line breaks

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 10]
```diff
@@ -0,0 +1,249 @@
...
+params = init_field.dofs, lagrange_params
+#This is just to initialize an empty state for the lagrange multiplier update and get some information
+lag_state,grad,info=ALM.init(params)
+
+
+
+
+
+
+
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Linebreaks

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py [hunk 11]
```diff
@@ -0,0 +1,249 @@
...
+opt_coils.plot(ax=ax2, show=False,label='Standard optimized coils')
+surface.plot(ax=ax2, show=False)
+ax3 = fig.add_subplot(133, projection='3d')
+opt_coils_alm.plot(ax=ax3, show=False,label='ALM optimized coils')
+surface.plot(ax=ax3, show=False)
+plt.legend()
+plt.tight_layout()
+plt.show()
+
+EXPORT = False
```
Comment (rogeriojorge) [inline examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Again, input parameters like these should be at the top

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
A few things:
- With a few changes the caching can be made more future-proof. The existing losses.py pattern clears cached DOFs when dependencies change. The new CompositeConstraint and SelectiveConstraint cache _starting_dofs / _dofs_to_pytree, but their dependencies setters do not clear those caches. If dependencies are reassigned after first access, the flattened DOFs can silently refer to old objects.
My take would be to reuse the losses.py cache/dependency pattern instead of reimplementing it. losses.py already has base_loss, custom_loss, dependency setters, starting_dofs, and dofs_to_pytree behavior. The new constraint classes duplicate much of that behavior.

- I didn't test it locally, but seems that default arguments for ALM_model_jaxopt_LevenbergMarquardt might fail, as it has model_mu='Mu_Tolerance', but the shown implementation only defines update_fn inside the if model_mu=='Mu_Adaptative_2' branch, then returns partial(update_fn, ...) always. That means the default path can hit an undefined local variable.

- The tests should follow the new api change of LagrangeMultiplier from 3 fields to 5 required fields: value, penalty, omega, eta, sq_grad

- eq() and ineq() repeatedly call fun(*args, **kwargs) just to build zeros_like arrays. For eq, the init path calls fun several times in one expression. For ineq, it computes out once, then still calls fun repeatedly for the multiplier fields. This can become:

def _multiplier_like(out, multiplier, penalty, omega, eta, sq_grad):
    z = jnp.zeros_like(out)
    return LagrangeMultiplier(
        value=multiplier + z,
        penalty=penalty + z,
        omega=omega + z,
        eta=eta + z,
        sq_grad=sq_grad + z,
    )

Then both eq and ineq init functions become much shorter and cheaper.

- Not now, but maybe in the future, we could refactor some of the code to make it easier to manage and easier to catch bugs. The Optax, JAXopt LBFGS-B, JAXopt LBFGS, JAXopt LM, and Optimistix LM paths all repeat the same pattern: build lagrangian, compute omega_min, run inner solver, recompute grad/info, update lagrange params, recompute grad/info again. You could extract:

def _make_lagrangian(constraints, loss, model_lagrangian):
    ...

def _omega_min(lagrange_params):
    ...

def _lagrange_updates(lagrange_params, grad, info, old_info, model_mu, model_lagrangian, ...):
    ...

def _finish_outer_step(main_params, lagrange_params, lag_updates, lagrangian, **kwargs):
    ...

That would reduce the duplicated branches.

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 11 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: I have addressed most comments. The only thing is that  I kept the different method as these will be addressed in future modification. So I think it is better to keep them and modifying after, than erase those extra methods and then add corrected versions later.

File: essos/augmented_lagrangian.py [hunk 12]
```diff
@@ -533,8 +1011,10 @@ def update_fn(params, lag_state,grad,info,eta,omega,beta=beta,mu_max=mu_max,alph
 
 
 #This uses jaxopt LevenbergMarquardt least squares (not working on jax==0.6.0 in GPU due to cuda versions conflict. Works on jax==0.5.0)
-def ALM_model_jaxopt_LevenbergMarquardt(constraints: Constraint,#List of constraints
+def ALM_model_jaxopt_LevenbergMarquardt(constraints: BaseConstraint,#List of constraints
     loss= lambda x: 0.,                    #function which represents the loss   (Callable, default 0.)
+    model_lagrangian='Squared',            #Model to use for updating lagrange multipliers
+    model_mu='Mu_Tolerance',               #Model to use for updating lagrange multipliers
```
Comment (rogeriojorge) [inline essos/augmented_lagrangian.py:1017]: The main thing I still see is in ALM_model_jaxopt_LevenbergMarquardt(): the default is now model_mu="Mu_Tolerance", but in the current diff update_fn is only defined inside the if model_mu == "Mu_Adaptative_2". Then the function returns partial(update_fn, ...) always. With the default path, that can still fail with update_fn undefined.

**Review by rogeriojorge — CHANGES_REQUESTED** [review]
While there are some still unresolved comments (too many comments, line breaks, parameters at the top, and optional metods for future) these are small. After the update_fn default path issue is resolved we can swiftly move forward for merge.

**Comment by eduardolneto** [comment]
Addressed last comment

File: essos/augmented_lagrangian.py (hunk 12 above)
Comment (eduardolneto) [reply essos/augmented_lagrangian.py:1017]: This is already addressed

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 11 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Done

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 10 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Done

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 9 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Done

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 8 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py:118]: Done

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 7 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py]: Done

File: examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py (hunk 6 above)
Comment (eduardolneto) [reply examples/coil_optimization/optimize_coils_vmec_surface_augmented_lagrangian_comparison.py:70]: Done

**Review by rogeriojorge — APPROVED** [review]
Looks good to me
